File: classifier/data/raw/human/samples/project_euler_zig/problem_037.zig
const std = @import("std");

// Project Euler 37: sum the eleven primes truncatable from both directions.
pub fn main() !void {
    const limit = 1_000_000;
    var prime = [_]bool{true} ** limit;
    prime[0] = false;
    prime[1] = false;
    var p: usize = 2;
    while (p * p < limit) : (p += 1) {
        if (prime[p]) {
            var multiple = p * p;
            while (multiple < limit) : (multiple += p) prime[multiple] = false;
        }
    }

    var found: u8 = 0;
    var sum: u32 = 0;
    var n: usize = 11;
    while (n < limit and found < 11) : (n += 1) {
        if (prime[n] and isTruncatable(n, &prime)) {
            found += 1;
            sum += @intCast(n);
        }
    }
    try printAnswer(sum);
}

fn isTruncatable(number: usize, prime: []const bool) bool {
    var right = number;
    while (right >= 10) {
        right /= 10;
        if (!prime[right]) return false;
    }

    var divisor: usize = 1;
    while (divisor <= number / 10) : (divisor *= 10) {}
    var left = number;
    while (divisor > 1) {
        left %= divisor;
        if (!prime[left]) return false;
        divisor /= 10;
    }
    return true;
}

fn printAnswer(value: anytype) !void {
    var buffer: [64]u8 = undefined;
    const text = try std.fmt.bufPrint(&buffer, "{d}\n", .{value});
    try std.Io.File.stdout().writeStreamingAll(std.Options.debug_io, text);
}

File: classifier/data/raw/ai/samples/project_euler_java/Problem037.java
/** Project Euler 37: sum of the eleven primes truncatable from left and right. */
public class Problem037 {
    public static void main(String[] args) {
        boolean[] prime = sieve(1_000_000);
        int count = 0;
        int sum = 0;
        for (int n = 11; count < 11; n += 2) {
            if (prime[n] && isTruncatable(n, prime)) {
                sum += n;
                count++;
            }
        }
        System.out.println(sum);
    }

    private static boolean isTruncatable(int n, boolean[] prime) {
        for (int right = n / 10; right > 0; right /= 10) {
            if (!prime[right]) {
                return false;
            }
        }
        int divisor = 1;
        while (divisor <= n / 10) {
            divisor *= 10;
        }
        for (int left = n; divisor > 1; divisor /= 10) {
            left %= divisor;
            if (!prime[left]) {
                return false;
            }
        }
        return true;
    }

    private static boolean[] sieve(int limit) {
        boolean[] prime = new boolean[limit];
        for (int i = 2; i < limit; i++) {
            prime[i] = true;
        }
        for (int i = 2; i * i < limit; i++) {
            if (prime[i]) {
                for (int multiple = i * i; multiple < limit; multiple += i) {
                    prime[multiple] = false;
                }
            }
        }
        return prime;
    }
}
